### ISSUE TITLE
Mock package in monorepo

### ISSUE THREAD

**Issue by fooddilsn** [vitest-dev/vitest#3265]
### Describe the bug

I am facing an issue mocking a package in a monorepo.

When I use `vi.mock('whatever')` without specifying an inline factory, I would expect vitest to look for the mock in the app's `__mocks__` folder in the monorepo, but if I mock a package of the monorepo, that does not happen.
I have added some logs [here](https://github.com/vitest-dev/vitest/blob/7c8f0ba989aaa3152f939d6d6adbf55a14e68473/packages/vitest/src/runtime/mocker.ts#L82) and [here](https://github.com/vitest-dev/vitest/blob/7c8f0ba989aaa3152f939d6d6adbf55a14e68473/packages/vitest/src/runtime/mocker.ts#L178) to gather more information.

When I mock a node module (e.g. undici), the output is as follows:

```
resolvePath {
  rawId: 'undici',
  id: '[...]/monorepo-vi-mock-resolver/node_modules/.pnpm/undici@5.22.0/node_modules/undici/index.js',
  fsPath: '[...]/monorepo-vi-mock-resolver/node_modules/.pnpm/undici@5.22.0/node_modules/undici/index.js'
}
const external = !isAbsolute(fsPath) || fsPath.includes("/node_modules/") ? rawId : null
{
  'isAbsolute(fsPath)': true,
  'fsPath.includes("/node_modules/")': true,
  external: 'undici'
}

resolveMockPath {
  mockPath: '[...]/monorepo-vi-mock-resolver/node_modules/.pnpm/undici@5.22.0/node_modules/undici/index.js',
  external: 'undici',
  path: 'undici',
  'isNodeBuiltin(mockPath)': false,
  'existsSync(mockPath)': true
}
=> enter the if (external || isNodeBuiltin(mockPath) || !existsSync(mockPath))
{
  mockDirname: '.',
  mockFolder: '[...]/monorepo-vi-mock-resolver/apps/foo/__mocks__',
  'existsSync(mockFolder)': true
}
__mocks__ content { files: [ 'sample.ts', 'undici.ts' ], baseOriginal: 'undici' }
return [...]/monorepo-vi-mock-resolver/apps/foo/__mocks__/undici.ts
```

If I mock a monorepo package instead, the output looks like this:

```
resolvePath {
  rawId: 'sample',
  id: '[...]/monorepo-vi-mock-resolver/packages/sample/dist/index.js',
  fsPath: '[...]/monorepo-vi-mock-resolver/packages/sample/dist/index.js'
}
const external = !isAbsolute(fsPath) || fsPath.includes("/node_modules/") ? rawId : null
{
  'isAbsolute(fsPath)': true,
  'fsPath.includes("/node_modules/")': false,
  external: null
}

resolveMockPath {
  mockPath: '[...]/monorepo-vi-mock-resolver/packages/sample/dist/index.js',
  external: null,
  path: '[...]/monorepo-vi-mock-resolver/packages/sample/dist/index.js',
  'isNodeBuiltin(mockPath)': false,
  'existsSync(mockPath)': true
}
=> skip the if (external || isNodeBuiltin(mockPath) || !existsSync(mockPath))
{
  dir: '[...]/monorepo-vi-mock-resolver/packages/sample/dist',
  baseId: 'index.js',
  fullPath: '[...]/monorepo-vi-mock-resolver/packages/sample/dist/__mocks__/index.js'
}
return null // because the mock file (fullPath) does not exist
```

And the final map is:

```
mockMap {
  '[...]/monorepo-vi-mock-resolver/apps/foo/src/index.test.ts' => {
    '[...]/monorepo-vi-mock-resolver/node_modules/.pnpm/undici@5.22.0/node_modules/undici/index.js': '[...]/monorepo-vi-mock-resolver/apps/foo/__mocks__/undici.ts',
    '[...]/monorepo-vi-mock-resolver/packages/sample/dist/index.js': null
  }
}
```

In the second case, since the `fsPath` is an absolute path and does not contain `/node_modules/`, the dependency is not marked as "external" and the mock file is searched for in the `dist/__mocks__` folder of the package itself.

Is this an intended behaviour?

### Reproduction

https://github.com/fooddilsn/monorepo-vi-mock-resolver

### System Info

```shell
System:
  OS: macOS 13.2.1
  CPU: (10) arm64 Apple M1 Max
  Memory: 72.38 MB / 32.00 GB
  Shell: 5.8.1 - /bin/zsh
Binaries:
  Node: 16.17.1 - ~/.nvm/versions/node/v16.17.1/bin/node
  Yarn: 1.22.19 - ~/.nvm/versions/node/v16.17.1/bin/yarn
  npm: 8.15.0 - ~/.nvm/versions/node/v16.17.1/bin/npm
Browsers:
  Chrome: 112.0.5615.137
  Safari: 16.3
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
This is expected, there is no way to know if path is a "package" or not

**Comment by fooddilsn** [comment]
So how can I properly mock a package in a monorepo?

Creating a `__mocks__` folder in the `src` of the package to have it in the `dist` folder after compilation seems "wrong" to me, doesn't it? What if I want to mock the same package in different ways in different apps?

**Comment by fox1t** [comment]
@sheremet-va I am also curious to understand how to mock a package that belongs to the monorepo in another package/app of the same monorepo. I need this functionality for an OSS project and want to write a clear guide for new contributors.

**Comment by fox1t** [comment]
@sheremet-va I am not sure I understood. Let me try to rephrase this to check it. 
So if I have a TypeScript monorepo, you suggest putting a `mock` folder (outside the src folder) in every package, like it happens when you mock stuff inside apps. If so, it means that Vitest should be able to pick the mocked version automatically. Is this the case? What bothers me is that the entry point of the package will be inside the `dist` folder and not inside the `src` one, but here we are, of course, "running" .ts files and not the compiled code.

**Comment by sheremet-va** [comment]
The other solution would be to have a separate option in Vitest that tells it where modules are located, I guess. Currently everything inside `node_modules` is a module and will be checked in the root folder:

https://github.com/vitest-dev/vitest/blob/31f835f15618b863b3ca04b0952a9b330f0fd0be/packages/vitest/src/runtime/mocker.ts#L86

I don't think we should reuse `workspaces` because you can use them for different sorts of things, and not only for package management.

**Comment by fox1t** [comment]
I don't like the implicit reuse of workspaces either, but I like the explicit configuration inside the Vitest config! 
It can be a map between the package name and its location inside the monorepo. @fooddilsn, would you like to submit a PR to implement this?

**Comment by sheremet-va** [comment]
> It can be a map between the package name and its location inside the monorepo.

I was thinking more about a list of folders like [in Jest](https://jestjs.io/docs/configuration#moduledirectories-arraystring).

**Comment by fooddilsn** [comment]
Sure, change the configuration is exactly what I was thinking about. What do you guys think about a new option related to the [`deps` config](https://github.com/vitest-dev/vitest/blob/main/docs/config/index.md#deps)? Something like `deps.monorepo: (string | RegExp)[]` that tells where the monorepo packages are?

**Comment by fox1t** [comment]
Folders Jest option seems more flexible and faster to implement. I would go with that one as the starting point. If needed, we can iterate over it in the future.
